File: frontend/src/api/creditCards.ts
import type {
  CreditCardCategoryDistributionPoint,
  CreditCardMetrics,
  CreditCardMonthlyActivityPoint,
  CreditCardStatementItem,
  CreditCardStatementUploadResponse
} from "../types/creditCard";

const API_BASE = "/api/v1/credit-cards";

async function fetchJson<T>(path: string, signal?: AbortSignal): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, {
    headers: { Accept: "application/json" },
    signal
  });
  if (!response.ok) {
    throw new Error(`Request failed (${response.status})`);
  }
  return response.json() as Promise<T>;
}

export function getCreditCardMetrics(signal?: AbortSignal): Promise<CreditCardMetrics> {
  return fetchJson<CreditCardMetrics>("/metrics", signal);
}

export function getCreditCardMonthlyActivity(
  signal?: AbortSignal
): Promise<CreditCardMonthlyActivityPoint[]> {
  return fetchJson<CreditCardMonthlyActivityPoint[]>("/monthly-activity", signal);
}

export function getCreditCardCategoryDistribution(
  signal?: AbortSignal
): Promise<CreditCardCategoryDistributionPoint[]> {
  return fetchJson<CreditCardCategoryDistributionPoint[]>("/category-distribution", signal);
}

export function getCreditCardStatementItems(
  signal?: AbortSignal
): Promise<CreditCardStatementItem[]> {
  return fetchJson<CreditCardStatementItem[]>("/statement-items", signal);
}

export async function uploadCreditCardStatement(
  file: File,
  signal?: AbortSignal
): Promise<CreditCardStatementUploadResponse> {
  const body = new FormData();
  body.append("file", file);
  const response = await fetch(`${API_BASE}/statements/upload`, {
    body,
    headers: { Accept: "application/json" },
    method: "POST",
    signal
  });
  if (!response.ok) {
    const payload = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(payload?.detail ?? `Request failed (${response.status})`);
  }
  return response.json() as Promise<CreditCardStatementUploadResponse>;
}

File: backend/app/routers/credit_cards.py
from fastapi import APIRouter, Depends, File, UploadFile
from sqlalchemy.orm import Session

from app.config import Settings, get_settings
from app.database import get_db
from app.repositories.credit_card_repo import CreditCardRepository
from app.schemas.credit_card import (
    CreditCardCategoryDistributionPoint,
    CreditCardMetrics,
    CreditCardMonthlyActivityPoint,
    CreditCardStatementItemResponse,
    CreditCardStatementUploadResponse,
)
from app.services.credit_card_analytics import CreditCardAnalyticsService
from app.services.credit_card_ingestion import CreditCardIngestionError, CreditCardIngestionService

router = APIRouter(prefix="/api/v1/credit-cards", tags=["credit-cards"])


def get_credit_card_repository(db: Session = Depends(get_db)) -> CreditCardRepository:
    return CreditCardRepository(db)


def get_credit_card_analytics(
    repository: CreditCardRepository = Depends(get_credit_card_repository),
) -> CreditCardAnalyticsService:
    return CreditCardAnalyticsService(repository)


@router.post("/statements/upload", response_model=CreditCardStatementUploadResponse)
async def upload_credit_card_statement(
    file: UploadFile = File(...),
    repository: CreditCardRepository = Depends(get_credit_card_repository),
    settings: Settings = Depends(get_settings),
) -> CreditCardStatementUploadResponse:
    try:
        return await CreditCardIngestionService(repository, settings).ingest_upload(file)
    except CreditCardIngestionError as exc:
        from fastapi import HTTPException

        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.get("/metrics", response_model=CreditCardMetrics)
def get_credit_card_metrics(
    analytics: CreditCardAnalyticsService = Depends(get_credit_card_analytics),
) -> CreditCardMetrics:
    return analytics.get_metrics()


@router.get("/monthly-activity", response_model=list[CreditCardMonthlyActivityPoint])
def get_credit_card_monthly_activity(
    analytics: CreditCardAnalyticsService = Depends(get_credit_card_analytics),
) -> list[CreditCardMonthlyActivityPoint]:
    return analytics.get_monthly_activity()


@router.get("/category-distribution", response_model=list[CreditCardCategoryDistributionPoint])
def get_credit_card_category_distribution(
    analytics: CreditCardAnalyticsService = Depends(get_credit_card_analytics),
) -> list[CreditCardCategoryDistributionPoint]:
    return analytics.get_category_distribution()


@router.get("/statement-items", response_model=list[CreditCardStatementItemResponse])
def get_credit_card_statement_items(
    analytics: CreditCardAnalyticsService = Depends(get_credit_card_analytics),
) -> list[CreditCardStatementItemResponse]:
    return analytics.get_statement_items()
